[RFC] SIP Fields (#1014)
elastic/ecs: rfcs/text/0008-sip-fields.md

| sip.sdp.video.description[]   | wildcard     | video 57126 RTP/AVP 8 101 |
| sip.sdp.video.port            | keyword      | 57126 |
| sip.sdp.video.format[]        | wildcard     | 8, 101 (ITU-T G.711 PCMA, DynamicRTP-Type-101) |
| sip.sdp.video.attributes[]    | wildcard     | 0 PCMU/8000, 101 telephone-event/8000, fmtp:101 0-16 |

## Usage

<!--
Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
-->
Typical implementations will utilize these fields to describe and normalize the various stages of a SIP/SDP based communcations mechanism.  Additional considerations including call analytics, fraud detection, troubleshooting, and threat detection have been identified as additional considerations.

## Source data

Source Data will come from packet/protocol analysis from endpoints (e.g. Packetbeat) or network observers (e.g. Zeek/Corelight & Suricata), logs from SIP Servers (e.g. Cisco Call Manager, Microsoft Lync), or logs from SIP-aware perimeter devices (e.g. Palo Alto NGFW).

> <!--

Reviewer: ```suggestion

See this example of [raw SIP header](0008/Sip-via-ordering-example.txt).

<!--
```

# 0000: SIP Fields
<!-- Leave this ID at 0000. The ECS team will assign a unique, contiguous RFC number upon merging the initial stage of this RFC. -->

> - Stage: **0 (strawperson)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->

Reviewer 2: Between the discussions we've had already and the tremendous work that @dainperkins has already invested, I suggest targeting stage 1 here.

What do others think?

Reviewer: Agreed with stage 1 from the start

> - Stage: **0 (strawperson)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->

Reviewer 2: ```suggestion
- Stage: **1 (proposal)** <!-- Update to reflect target stage. See https://elastic.github.io/ecs/stages.html -->
```

> - Date: **TBD** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->

Reviewer: ```suggestion
- Date: **2020-12-04** <!-- The ECS team sets this date at merge time. This is the date of the latest stage advancement. -->
```

<!--
As you work on your RFC, use the "Stage N" comments to guide you in what you should focus on, for the stage you're targeting.
Feel free to remove these comments as you go along.
-->

<!--
Stage 0: Provide a high level summary of the premise of these changes. Briefly describe the nature, purpose, and impact of the changes. ~2-5 sentences.
-->

ECS SIP Fields provide normalization for fields related to Session Initiation and Session Description Protocols used in IP based real time communications (voice, video, sip based messaging).

## Fields

| Sip Fields                	| Example                                      	| type      	|
|---------------------------	|----------------------------------------------	|-----------	|
|                           	|                                              	|           	|
| **SIP Request**               | INVITE sip:test@10.0.2.15:5060 SIP/2.0       	|           	|
| sip.type                  	| request / response                           	| keyword   	|
| sip.method                	| invite                                       	| keyword   	|

> | sip.uri.original          	| test@10.0.2.15:5060                          	| keyword   	|

Author: did it for all of the uri fields, and removed the text multifield.

| sip.uri.port              	| 5060                                         	| long      	|
| sip.version               	| 2                                            	| keyword   	|
|                           	|                                              	|           	|
| **SIP Response**              | SIP/2.0 200 OK                               	|           	|
| sip.code                  	| 200                                          	| keyword   	|
| sip.status                	| ok                                           	| keyword   	|
| sip.version               	| 2                                            	| keyword   	|
|                           	|                                              	|           	|
| **SIP Headers**               |                                              	|           	|
| sip.accept                	| application/sdp                              	| keyword   	|
| sip.allow[]               	| REGISTER, INVITE, ACK, BYE                   	| keyword[] 	|
| sip.call_id               	| 1-1966@10.0.2.20                             	| keyword   	|

> | sip.content_length        	| 0                                            	| integer   	|

Author: not really...per packet integer would work - per 1m flow or conection might not

| sip.private.uri.original  	| sip:35104723@sip.cybercity.dk                	| keyword   	|
| sip.private.original.text 	| sip.cybercity.dk                             	| text      	|
| sip.private.uri.scheme    	| sip                                          	| keyword   	|
| sip.private.username      	| 35104723                                     	| keyword   	|
| sip.supported[]           	| timer, path, replaces                        	| keyword[] 	|
| user_agent.original       	| FreeSWITCH-mod_sofia/1.6.12-20-b91a0a6~64bit 	| keyword   	|
| user_agent.original.text  	| FreeSWITCH-mod_sofia/1.6.12-20-b91a0a6~64bit 	| text      	|
|                           	|                                              	|           	|
| **SIP Headers CSEQ**        	| 68 invite                                    	|           	|
| sip.cseq.code             	| 68                                           	| integer   	|
| sip.cseq.method           	| invite                                       	| keyword   	|
|                               |                                               |               |

> | **SIP Headers Via**              | Ex 1: SIP/2.0/UDP 192.168.1.2;received=80.230.219.70;rport=5060;branch=z9hG4bKnp112903503-43a64480192.168.1.2 EX2: SIP/2.0/UDP 10.0.2.20:5060;branch=z9hG4bK-1966-1-0                                               	|             	|

Author: yeah theres a possibility of multiple proxies in the mix.  ordering would be nice, but I'm unsure what it looks like

> | **SIP Headers Via**              | Ex 1: SIP/2.0/UDP 192.168.1.2;received=80.230.219.70;rport=5060;branch=z9hG4bKnp112903503-43a64480192.168.1.2 EX2: SIP/2.0/UDP 10.0.2.20:5060;branch=z9hG4bK-1966-1-0                                               	|             	|

Reviewer 3: examples of such bellow. ordering matters. The topmost via is the most recent along the source-destination path. Some other headers may also occur multiple times, either as multiple headers, or multiple header values, or combination of both: `Record-Route` and `Route` often. 

```
U 192.168.0.85:5060 -> 192.168.0.155:6060
INVITE sip:5c47b8a83af57797@192.168.0.155:6060;transport=udp SIP/2.0.
Record-Route: <sip:192.168.0.85;ftag=1B8F4BD4-5FA0B47F000630A1-E25B7700;avp=Jr0DBwBhY2NvdW50AwB5ZXMDBgBzdGltZXICADkwAwkAZGlhbG9nX2lkFgAwNGQzLTVmMjNkNDk2LTMxYTgzODNj;lr=on>.
Via: SIP/2.0/UDP 192.168.0.85;branch=z9hG4bK94b5.5043d304.0.
Via: SIP/2.0/UDP 192.168.0.155:6060;branch=z9hG4bK~tW~Hade;rport=6060.
From: "jiri" <sip:jiri@iptel.org>;tag=1B8F4BD4-5FA0B47F000630A1-E25B7700.
To: <sip:103@frafos.net>.
CSeq: 11 INVITE.
Call-ID: 38E5D275-5FA0B47F000630AB-E25B7700.
Max-Forwards: 4.
Supported: ice, outbound, replaces, timer.
Session-Expires: 90.
Allow: INVITE,ACK,CANCEL,BYE,UPDATE,MESSAGE,OPTIONS,REFER,INFO.
User-Agent: JsSIP 3.0.12.
X-Org-ConnID: 94.142.239.227-20201103023807-cMigNfdo+tg2VUv.
X-Call-ID: qfg0157hu3mujhmuufii.
Content-Type: application/sdp.
Content-Length: 4642.
Contact: <sip:1B8F4BD4-5FA0B47F000630A1-E25B7700@192.168.0.155:6060;transport=udp>.
P-hint: usrloc applied.
P-hint-DBG: forward-sbc.
.
v=0.
o=- 1185881956307229883 2 IN IP4 192.168.0.155.
s=-.
...
```

| sip.via.transport             	| udp                                                                                                                                                                                                                 	| keyword     	|
| sip.via.sent_by.address       	| 192.168.1.2                                                                                                                                                                                                         	| keyword     	|
| sip.via.sent_by.port          	|                                                                                                                                                                                                                     	| long        	|
| sip.via.received.address      	| 80.230.219.70                                                                                                                                                                                                       	| keyword     	|

> | sip..via.rport                	| 5060                                                                                                                                                                                                                	|             	|

Author: I think that's a good idea - I keep getting wrapped around the axle looking @ tcp headers (syn, ack, etc.) and icmp from the standpoint of unidirectional vs bidirectional src/dst... I think a larger discussion would be a great idea.

| sip.auth.uri.original         	| sip:192.168.1.2                                                                                                                                                                                                     	| keyword     	|
| sip.auth.uri.original.text    	| sip:192.168.1.2                                                                                                                                                                                                     	| text        	|
| sip.auth.uri.scheme           	| sip                                                                                                                                                                                                                 	| keyword     	|
| sip.auth.uri.host             	| 192.168.1.2                                                                                                                                                                                                         	| keyword     	|

> | sip.auth.uri.port             	| n/a                                                                                                                                                                                                                 	| long        	|

Author: No, just that thee example didn't list one - I left it blank in the update

| sip.sdp.video.format[]        	| 8, 101 (ITU-T G.711 PCMA, DynamicRTP-Type-101)                                                                                                                                                                      	| wildcard?[] 	|
| sip.sdp.video.attributes[]    	| 0 PCMU/8000, 101 telephone-event/8000, fmtp:101 0-16                                                                                                                                                                	| wildcard?[] 	|

## Usage

<!--
Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
-->
Typical implementations will utilize these fields to describe and normalize the various stages of a SIP/SDP based communcations mechanism.  Additional considerations including call analytics, fraud detection, troubleshooting, and threat detection have been identified as additional considerations.

## Source data

> Source Data will come from packet/protocol analysis from endpoints (e.g. Packetbeat) or network observers (e.g. Zeek/Corelight & Suricata), logs from SIP Servers (e.g. Cisco Call Manager, Microsfot Lync), or logs SiP aware perimeter devices (e.g. Palo Alto NGFW).

Reviewer 2: ```suggestion
Source Data will come from packet/protocol analysis from endpoints (e.g. Packetbeat) or network observers (e.g. Zeek/Corelight & Suricata), logs from SIP Servers (e.g. Cisco Call Manager, Microsfot Lync), or logs from SIP-aware perimeter devices (e.g. Palo Alto NGFW).
```

<!--
Stage 1: Provide a high-level description of example sources of data. This does not yet need to be a concrete example of a source document, but instead can simply describe a potential source (e.g. nginx access log). This will ultimately be fleshed out to include literal source examples in a future stage. The goal here is to identify practical sources for these fields in the real world. ~1-3 sentences or unordered list.
-->

<!--
Stage 2: Included a real world example source document. Ideally this example comes from the source(s) identified in stage 1. If not, it should replace them. The goal here is to validate the utility of these field changes in the context of a real world example. Format with the source name as a ### header and the example document in a GitHub code block with json formatting.
-->

<!--
Stage 3: Add more real world example source documents so we have at least 2 total, but ideally 3. Format as described in stage 2.
-->

## Scope of impact

> No impact expected as SIP fieldsets are new, and will not impact any existing fields.

Reviewer 2: It looks like yes, the Packetbeat SIP/SDP support is out in 7.10 as beta ([reference](https://www.elastic.co/guide/en/beats/packetbeat/current/configuration-protocols.html)).

<!--
Stage 2: Identifies scope of impact of changes. Are breaking changes required? Should deprecation strategies be adopted? Will significant refactoring be involved? Break the impact down into:
 * Ingestion mechanisms (e.g. beats/logstash)
 * Usage mechanisms (e.g. Kibana applications, detections)
 * ECS project (e.g. docs, tooling)
The goal here is to research and understand the impact of these changes on users in the community and development teams across Elastic. 2-5 sentences each.
-->

## Concerns

> Normalization, and the degree of normalization, of SIP URI fields may be an issue for discussion based on the potential implementation of ingesting SIP call records for the purposes of review for e.g. various types of communications fraud (e.g. should PSTN numbers be normalized with international dial codes, should implementaiton include capabilities to define internal call plans for more effective analysis, etc.)

Author: fixed

> Normalization of SIP/SDP and real time communication protocol connections may require the definition of a field similar to network.community_id to allow for the trackign of the full scope of a connection. Additionally the ininital SDP setup phase often includes

Reviewer 2: ```suggestion
Normalization of SIP/SDP and real time communication protocol connections may require the definition of a field similar to network.community_id to allow for the tracking of the full scope of a connection. Additionally the initial SDP setup phase often includes
```

<!--
Stage 1: Identify potential concerns, implementation challenges, or complexity. Spend some time on this. Play devil's advocate. Try to identify the sort of non-obvious challenges that tend to surface later. The goal here is to surface risks early, allow everyone the time to work through them, and ultimately document resolution for posterity's sake.
-->

<!--
Stage 2: Document new concerns or resolutions to previously listed concerns. It's not critical that all concerns have resolutions at this point, but it would be helpful if resolutions were taking shape for the most significant concerns.
-->

<!--
Stage 3: Document resolutions for all existing concerns. Any new concerns should be documented along with their resolution. The goal here is to eliminate the risk of churn and instability by resolving outstanding concerns.
-->

<!--
Stage 4: Document any new concerns and their resolution. The goal here is to eliminate risk of churn and instability by ensuring all concerns have been addressed.
-->

## Real-world implementations

Packetbeat Implementation (packet/protocol analysis)
Zeek Implementation (packet/protocol/connection analysis)
Cisco Call Manager (Log ingestion)

> Mictosoft Lync (Log Analysis)

Reviewer 2: ```suggestion
Microsoft Lync (Log Analysis)
```

> Microsoft Lync (Log Analysis)

Reviewer: Here we're looking to list the real world implementation of using the SIP ECS fields. You can keep Packetbeat here, since the goal is for it to be adjusted to match the ECS fields.

The other ones should be removed for now, though. They're already correctly mentioned in the "data sources" section.

Note that when the Zeek SIP parser is adjusted to follow ECS we can list it here as well. But hopefully it won't be adjusted until this RFC reaches at least stage 2, when fields will be more stable.

Author: commented out

<!--
Stage 4: Identify at least one real-world, production-ready implementation that uses these updated field definitions. An example of this might be a GA feature in an Elastic application in Kibana.
-->

## People

The following are the people that consulted on the contents of this RFC.

> * @DainPerkins | Author

Reviewer: @jamiehynds just discussed with @marc-gr, and Marc will the sponsor for this RFC. Could you add him here as well please?

> * @DainPerkins | Author

Author: added both

<!--
Who will be or has been consulted on the contents of this RFC? Identify authorship and sponsorship, and optionally identify the nature of involvement of others. Link to GitHub aliases where possible. This list will likely change or grow stage after stage.

e.g.:

* @Yasmina | author
* @Monique | sponsor
* @EunJung | subject matter expert
* @JaneDoe | grammar, spelling, prose
* @Mariana
-->

## References

<!-- Insert any links appropriate to this RFC in this section. -->
https://tools.ietf.org/html/rfc3261
https://tools.ietf.org/html/rfc5621
https://tools.ietf.org/html/rfc5630
https://tools.ietf.org/html/rfc6878
https://tools.ietf.org/html/rfc8591
https://tools.ietf.org/id/draft-ietf-sipclf-format-05.html
https://www.sipforum.org/

https://github.com/elastic/ecs/issues/420

### RFC Pull Requests

<!-- An RFC should link to the PRs for each of it stage advancements. -->

> * Stage 0: https://github.com/elastic/ecs/pull/NNN

Reviewer 2: ```suggestion
* Stage 0: https://github.com/elastic/ecs/pull/1014
```
